src/include/sof/ipc/common.h

```diff
@@ -23,7 +23,9 @@ struct dma_sg_elem_array;
 struct ipc_msg;
 
 /* generic IPC header regardless of ABI MAJOR type that is always 4 byte aligned */
-typedef uint32_t ipc_cmd_hdr;
+typedef struct ipc_cmd_hdr_ {
+	uint32_t dat[2];
+} ipc_cmd_hdr;
```

Comment:
This should be opaque and abstract - can we make this a generic pointer instead (and then cast appropriately in IPC3/4) ?

Reply:
Hi Liam,

The generic pointer (uint32_t *) as used previously is what I'm trying to fix in this patch among other issues.
There are spots in code that do the casting back and forth between this and the larger IPC structures and then access by indexing with respect to this pointer which is an unsafe practice. I'm trying to make the ipc solution more robust and error proof for the future by this.

BR, Rafal

Reply:
Sorry, I mean we could do something generic/opaque like
```c
struct ipc_cmd_hdr; /* opaque type in header for declarations, defined in IPC3/4 handler code */
```

Reply:
Hi @lgirdwood. I have done the changes to abstract the ipc_cmd_hdr for ipc3 / ipc4 according to your suggestion (had to do it in ipc*/header.h and not in ipc*/handler.c because the definition is also used outside of handler.c).
The thing is however that [checktree (IPC_VERSION=IPC4 UNSIGNED_list= SIGNED_list='tgl tgl-h cnl icl jsl')](https://github.com/thesofproject/sof/runs/6224422069?check_suite_focus=true#logs)
tries to build e.g. cavs/platform.c using ipc4.
Effectively in the tree we have a combination of a platform and different MAJOR ABI (e.g. cavs + IPC3 and cavs + IPC4) which brings me to conclusion that ipc_cmd_hdr definition should in fact be common for IPC3/IPC4 to keep the platform code happy.
Do you have other suggestion in this case?

Reply:
> Effectively in the tree we have a combination of a platform and different MAJOR ABI (e.g. cavs + IPC3 and cavs + IPC4) which brings me to conclusion that ipc_cmd_hdr definition should in fact be common for IPC3/IPC4 to keep the platform code happy.
> Do you have other suggestion in this case?

So IIUC, we have a generic type for header that we define in IPC3 and IPC4 code which is fine. We could also define this in the IPC driver logic to align with the HW (and not the IPC flavour), meaning the IPC drivers will always build regardless of any IPC version (including any future IPC updates).